src/sql/src/statements/option_map.rs

```diff
@@ -12,52 +12,142 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-mod visit;
-mod visit_mut;
-
-use std::borrow::Borrow;
 use std::collections::HashMap;
-use std::iter::FromIterator;
+use std::ops::ControlFlow;
+
+use common_base::secrets::{ExposeSecret, ExposeSecretMut, SecretString};
+use itertools::Itertools;
+use sqlparser::ast::{Visit, VisitMut, Visitor, VisitorMut};
+
+const REDACTED_OPTIONS: [&str; 2] = ["access_key_id", "secret_access_key"];
 
 /// Options hashmap.
-/// Because the trait `Visit` and `VisitMut` is not implemented for `HashMap<String, String>`, we have to wrap it and implement them by ourself.
-#[derive(Clone, Eq, PartialEq, Debug)]
+#[derive(Clone, Debug, Default)]
 pub struct OptionMap {
-    pub map: HashMap<String, String>,
+    options: HashMap<String, String>,
+    secrets: HashMap<String, SecretString>,
 }
 
 impl OptionMap {
     pub fn insert(&mut self, k: String, v: String) {
-        self.map.insert(k, v);
+        if REDACTED_OPTIONS.contains(&k.as_str()) {
+            self.secrets.insert(k, SecretString::new(Box::new(v)));
+        } else {
+            self.options.insert(k, v);
+        }
     }
 
     pub fn get(&self, k: &str) -> Option<&String> {
-        self.map.get(k)
+        if let Some(value) = self.options.get(k) {
+            Some(value)
+        } else if let Some(value) = self.secrets.get(k) {
+            Some(value.expose_secret())
+        } else {
+            None
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.options.is_empty() && self.secrets.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.options.len() + self.secrets.len()
+    }
+
+    pub fn to_str_map(&self) -> HashMap<&str, &str> {
+        let mut map = HashMap::new();
+        map.extend(self.options.iter().map(|(k, v)| (k.as_str(), v.as_str())));
+        map.extend(
+            self.secrets
+                .iter()
+                .map(|(k, v)| (k.as_str(), v.expose_secret().as_str())),
+        );
+        map
+    }
+
+    pub fn into_map(self) -> HashMap<String, String> {
+        let mut map = self.options;
+        map.extend(
+            self.secrets
+                .into_iter()
+                .map(|(k, v)| (k, v.expose_secret().to_string())),
+        );
+        map
+    }
+
+    pub fn kv_pairs(&self) -> Vec<String> {
+        let mut result = Vec::with_capacity(self.options.len() + self.secrets.len());
+        for key in self.options.keys().sorted() {
```

Comment:
Is it necessary to sort the options?

Reply:
This is asked and the same reason as https://github.com/GreptimeTeam/greptimedb/pull/3744#discussion_r1577214798.

Perhaps we can use `BTreeMap` or other stable iterable collections and avoid this sort. But it doesn't affect too much IMO.

Reply:
If it's for testing purpose, you can sort and compare the final `Vec<String>`, in testing codes, outside this method.

Reply:
Many of our testing code compare the format result. I'd try to implement `Display` for `OptionMap` but still keep the sort manner. Generally we don't have more than 100 options and this overhead is minor.

Reply:
Use BTreeMap for a quick "fix".